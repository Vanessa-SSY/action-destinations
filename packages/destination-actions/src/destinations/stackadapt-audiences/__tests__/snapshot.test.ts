import { createTestEvent, createTestIntegration } from '@segment/actions-core'
import { generateTestData } from '../../../lib/test-data'
import destination from '../index'
import nock from 'nock'

const testDestination = createTestIntegration(destination)
const destinationSlug = 'actions-stackadapt-audiences'

describe(`Testing snapshot for ${destinationSlug} destination:`, () => {
  for (const actionSlug in destination.actions) {
    it(`${actionSlug} action - required fields`, async () => {
      const seedName = `${destinationSlug}#${actionSlug}`
      const action = destination.actions[actionSlug]
      const [eventData, settingsData] = generateTestData(seedName, destination, action, true)

      nock(/.*/).persist().get(/.*/).reply(200)
      nock(/.*/).persist().post(/.*/).reply(200)
      nock(/.*/).persist().put(/.*/).reply(200)

      const event = createTestEvent({
        properties: eventData,
      })
      event.properties!.marketing_status = 'Indeterminate'

      // Add required test data for specific actions
      let mapping = event.properties
      if (actionSlug === 'forwardAudienceEvent') {
        mapping = { ...event.properties, marketing_status: 'Indeterminate' }
      } else if (actionSlug === 'forwardProfile') {
        // Ensure forwardProfile has traits to avoid empty processing
        mapping = { 
          ...event.properties, 
          traits: { 
            email: 'test@example.com',
            firstName: 'Test',
            lastName: 'User',
            ...event.properties.traits 
          }
        }
      }

      const responses = await testDestination.testAction(actionSlug, {
        event: event,
        mapping: mapping,
        settings: settingsData,
        auth: undefined
      })

      if (!responses || responses.length === 0) {
        throw new Error(`No responses returned for action ${actionSlug}`)
      }

      const request = responses[0].request
      const rawBody = await request.text()

      try {
        const json = JSON.parse(rawBody)
        expect(json).toMatchSnapshot()
        return
      } catch (err) {
        expect(rawBody).toMatchSnapshot()
      }

      expect(request.headers).toMatchSnapshot()
    })

    it(`${actionSlug} action - all fields`, async () => {
      const seedName = `${destinationSlug}#${actionSlug}`
      const action = destination.actions[actionSlug]
      const [eventData, settingsData] = generateTestData(seedName, destination, action, false)

      nock(/.*/).persist().get(/.*/).reply(200)
      nock(/.*/).persist().post(/.*/).reply(200)
      nock(/.*/).persist().put(/.*/).reply(200)

      const event = createTestEvent({
        properties: eventData
      })

      // Add required test data for specific actions
      let mapping = event.properties
      if (actionSlug === 'forwardAudienceEvent') {
        mapping = { ...event.properties, marketing_status: 'Indeterminate' }
      } else if (actionSlug === 'forwardProfile') {
        // Ensure forwardProfile has traits to avoid empty processing
        mapping = { 
          ...event.properties, 
          traits: { 
            email: 'test@example.com',
            firstName: 'Test',
            lastName: 'User',
            ...event.properties.traits 
          }
        }
      }

      const responses = await testDestination.testAction(actionSlug, {
        event: event,
        mapping: mapping,
        settings: settingsData,
        auth: undefined
      })

      if (!responses || responses.length === 0) {
        throw new Error(`No responses returned for action ${actionSlug}`)
      }

      const request = responses[0].request
      const rawBody = await request.text()

      try {
        const json = JSON.parse(rawBody)
        expect(json).toMatchSnapshot()
        return
      } catch (err) {
        expect(rawBody).toMatchSnapshot()
      }
    })
  }
})