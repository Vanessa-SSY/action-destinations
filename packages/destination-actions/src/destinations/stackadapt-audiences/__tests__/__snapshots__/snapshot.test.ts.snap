// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Testing snapshot for actions-stackadapt-audiences destination: forwardAudienceEvent action - all fields 1`] = `
{
  "query": "mutation {
      upsertProfiles(
        input: {
          advertiserId: 84GW[vK%wv2xv@UF5iy,
          externalProvider: "segment_io",
          syncId: "5bcf6f4010d5fadefa88a2af6fd32324a30324a9ab69205f7c0df31c6f4ba012",
          profiles: "[{\\"userId\\":\\"84GW[vK%wv2xv@UF5iy\\",\\"email\\":\\"kagaca@til.zw\\",\\"first_name\\":\\"John\\",\\"last_name\\":\\"Doe\\",\\"phone\\":\\"1234567890\\",\\"address\\":\\"10 pleasant st.\\",\\"city\\":\\"Pleasantville\\",\\"state\\":\\"CA\\",\\"country\\":\\"USA\\",\\"postal_code\\":\\"12345\\",\\"timezone\\":\\"PST\\",\\"birth_day\\":23,\\"birth_month\\":3,\\"birth_year\\":1990,\\"birth_date\\":\\"1990-03-24\\",\\"testType\\":\\"84GW[vK%wv2xv@UF5iy\\",\\"audienceId\\":\\"84GW[vK%wv2xv@UF5iy\\",\\"audienceName\\":\\"84GW[vK%wv2xv@UF5iy\\",\\"action\\":\\"exit\\"}]"
        }
      ) {
        userErrors {
          message
        }
      }
      upsertProfileMapping(
        input: {
          advertiserId: 84GW[vK%wv2xv@UF5iy,
          mappingSchemaV2: [{incomingKey:"user_id",destinationKey:"external_id",label:"User ID",type:STRING,isPii:false},{incomingKey:"email",destinationKey:"email",label:"Email",type:STRING,isPii:true},{incomingKey:"first_name",destinationKey:"first_name",label:"First Name",type:STRING,isPii:true},{incomingKey:"last_name",destinationKey:"last_name",label:"Last Name",type:STRING,isPii:true},{incomingKey:"phone",destinationKey:"phone",label:"Phone",type:STRING,isPii:true},{incomingKey:"address",destinationKey:"address",label:"Address",type:STRING,isPii:true},{incomingKey:"city",destinationKey:"city",label:"City",type:STRING,isPii:false},{incomingKey:"state",destinationKey:"state",label:"State",type:STRING,isPii:false},{incomingKey:"country",destinationKey:"country",label:"Country",type:STRING,isPii:false},{incomingKey:"postal_code",destinationKey:"postal_code",label:"Postal Code",type:STRING,isPii:false},{incomingKey:"timezone",destinationKey:"timezone",label:"Timezone",type:STRING,isPii:false},{incomingKey:"birth_day",destinationKey:"birth_day",label:"Birth Day",type:NUMBER,isPii:false},{incomingKey:"birth_month",destinationKey:"birth_month",label:"Birth Month",type:NUMBER,isPii:false},{incomingKey:"birth_year",destinationKey:"birth_year",label:"Birth Year",type:NUMBER,isPii:true},{incomingKey:"birth_date",destinationKey:"birth_date",label:"Birth Date",type:STRING,isPii:true},{incomingKey:"testType",destinationKey:"testType",label:"TestType",type:STRING,isPii:false}],
          isOptedIn: false,
          mappableType: "segment_io"
        }
      ) {
        userErrors {
          message
        }
      }
      upsertExternalAudienceMapping(
          input: {
            advertiserId: 84GW[vK%wv2xv@UF5iy,
            mappingSchema: [{incomingKey:"audienceId",destinationKey:"external_id",type:STRING,label:"External Audience ID",isPii:false},{incomingKey:"audienceName",destinationKey:"name",type:STRING,label:"External Audience Name",isPii:false}],
            mappableType: "segment_io"
          }
        ) {
          userErrors {
            message
          }
        }
  }",
}
`;

exports[`Testing snapshot for actions-stackadapt-audiences destination: forwardAudienceEvent action - required fields 1`] = `
{
  "query": "mutation {
      upsertProfiles(
        input: {
          advertiserId: 84GW[vK%wv2xv@UF5iy,
          externalProvider: "segment_io",
          syncId: "0ec010cfb0502669fed268b6ae865dec38275f7e12292105e4047d9573a2e0f4",
          profiles: "[{\\"userId\\":\\"84GW[vK%wv2xv@UF5iy\\",\\"email\\":\\"kagaca@til.zw\\",\\"first_name\\":\\"John\\",\\"last_name\\":\\"Doe\\",\\"phone\\":\\"1234567890\\",\\"address\\":\\"10 pleasant st.\\",\\"city\\":\\"Pleasantville\\",\\"state\\":\\"CA\\",\\"country\\":\\"USA\\",\\"postal_code\\":\\"12345\\",\\"timezone\\":\\"PST\\",\\"birth_day\\":23,\\"birth_month\\":3,\\"birth_year\\":1990,\\"birth_date\\":\\"1990-03-24\\",\\"audienceId\\":\\"84GW[vK%wv2xv@UF5iy\\",\\"audienceName\\":\\"84GW[vK%wv2xv@UF5iy\\",\\"action\\":\\"exit\\"}]"
        }
      ) {
        userErrors {
          message
        }
      }
      upsertProfileMapping(
        input: {
          advertiserId: 84GW[vK%wv2xv@UF5iy,
          mappingSchemaV2: [{incomingKey:"user_id",destinationKey:"external_id",label:"User ID",type:STRING,isPii:false},{incomingKey:"email",destinationKey:"email",label:"Email",type:STRING,isPii:true},{incomingKey:"first_name",destinationKey:"first_name",label:"First Name",type:STRING,isPii:true},{incomingKey:"last_name",destinationKey:"last_name",label:"Last Name",type:STRING,isPii:true},{incomingKey:"phone",destinationKey:"phone",label:"Phone",type:STRING,isPii:true},{incomingKey:"address",destinationKey:"address",label:"Address",type:STRING,isPii:true},{incomingKey:"city",destinationKey:"city",label:"City",type:STRING,isPii:false},{incomingKey:"state",destinationKey:"state",label:"State",type:STRING,isPii:false},{incomingKey:"country",destinationKey:"country",label:"Country",type:STRING,isPii:false},{incomingKey:"postal_code",destinationKey:"postal_code",label:"Postal Code",type:STRING,isPii:false},{incomingKey:"timezone",destinationKey:"timezone",label:"Timezone",type:STRING,isPii:false},{incomingKey:"birth_day",destinationKey:"birth_day",label:"Birth Day",type:NUMBER,isPii:false},{incomingKey:"birth_month",destinationKey:"birth_month",label:"Birth Month",type:NUMBER,isPii:false},{incomingKey:"birth_year",destinationKey:"birth_year",label:"Birth Year",type:NUMBER,isPii:true},{incomingKey:"birth_date",destinationKey:"birth_date",label:"Birth Date",type:STRING,isPii:true}],
          isOptedIn: false,
          mappableType: "segment_io"
        }
      ) {
        userErrors {
          message
        }
      }
      upsertExternalAudienceMapping(
          input: {
            advertiserId: 84GW[vK%wv2xv@UF5iy,
            mappingSchema: [{incomingKey:"audienceId",destinationKey:"external_id",type:STRING,label:"External Audience ID",isPii:false},{incomingKey:"audienceName",destinationKey:"name",type:STRING,label:"External Audience Name",isPii:false}],
            mappableType: "segment_io"
          }
        ) {
          userErrors {
            message
          }
        }
  }",
}
`;

exports[`Testing snapshot for actions-stackadapt-audiences destination: forwardProfile action - all fields 1`] = `
{
  "query": "mutation {
      upsertProfiles(
        input: {
          advertiserId: PsAwlRv%,
          externalProvider: "segment_io",
          syncId: "20b72c0b0e460d68c78055bf266f8de715700c84d7e67cd2ba0c3c59254315a3",
          profiles: "[{\\"userId\\":\\"PsAwlRv%\\",\\"email\\":\\"bobgolvij@relsifhu.ag\\",\\"first_name\\":\\"John\\",\\"last_name\\":\\"Doe\\",\\"phone\\":\\"1234567890\\",\\"address\\":\\"10 pleasant st.\\",\\"city\\":\\"Pleasantville\\",\\"state\\":\\"CA\\",\\"country\\":\\"USA\\",\\"postal_code\\":\\"12345\\",\\"timezone\\":\\"PST\\",\\"birth_day\\":23,\\"birth_month\\":3,\\"birth_year\\":1990,\\"birth_date\\":\\"1990-03-24\\",\\"testType\\":\\"PsAwlRv%\\",\\"previous_id\\":\\"PsAwlRv%\\"}]"
        }
      ) {
        userErrors {
          message
        }
      }
      upsertProfileMapping(
        input: {
          advertiserId: PsAwlRv%,
          mappingSchemaV2: [{incomingKey:"user_id",destinationKey:"external_id",label:"User ID",type:STRING,isPii:false},{incomingKey:"email",destinationKey:"email",label:"Email",type:STRING,isPii:true},{incomingKey:"first_name",destinationKey:"first_name",label:"First Name",type:STRING,isPii:true},{incomingKey:"last_name",destinationKey:"last_name",label:"Last Name",type:STRING,isPii:true},{incomingKey:"phone",destinationKey:"phone",label:"Phone",type:STRING,isPii:true},{incomingKey:"address",destinationKey:"address",label:"Address",type:STRING,isPii:true},{incomingKey:"city",destinationKey:"city",label:"City",type:STRING,isPii:false},{incomingKey:"state",destinationKey:"state",label:"State",type:STRING,isPii:false},{incomingKey:"country",destinationKey:"country",label:"Country",type:STRING,isPii:false},{incomingKey:"postal_code",destinationKey:"postal_code",label:"Postal Code",type:STRING,isPii:false},{incomingKey:"timezone",destinationKey:"timezone",label:"Timezone",type:STRING,isPii:false},{incomingKey:"birth_day",destinationKey:"birth_day",label:"Birth Day",type:NUMBER,isPii:false},{incomingKey:"birth_month",destinationKey:"birth_month",label:"Birth Month",type:NUMBER,isPii:false},{incomingKey:"birth_year",destinationKey:"birth_year",label:"Birth Year",type:NUMBER,isPii:true},{incomingKey:"birth_date",destinationKey:"birth_date",label:"Birth Date",type:STRING,isPii:true},{incomingKey:"testType",destinationKey:"testType",label:"TestType",type:STRING,isPii:false}],
          isOptedIn: true,
          mappableType: "segment_io"
        }
      ) {
        userErrors {
          message
        }
      }
      
  }",
}
`;

exports[`Testing snapshot for actions-stackadapt-audiences destination: forwardProfile action - required fields 1`] = `
{
  "query": "mutation {
      upsertProfiles(
        input: {
          advertiserId: PsAwlRv%,
          externalProvider: "segment_io",
          syncId: "4716cb456cac7b37f63da0d042816219b43fa1f45de9e140cd38c58114ba4b0a",
          profiles: "[{\\"userId\\":\\"PsAwlRv%\\",\\"email\\":\\"bobgolvij@relsifhu.ag\\",\\"first_name\\":\\"John\\",\\"last_name\\":\\"Doe\\",\\"phone\\":\\"1234567890\\",\\"address\\":\\"10 pleasant st.\\",\\"city\\":\\"Pleasantville\\",\\"state\\":\\"CA\\",\\"country\\":\\"USA\\",\\"postal_code\\":\\"12345\\",\\"timezone\\":\\"PST\\",\\"birth_day\\":23,\\"birth_month\\":3,\\"birth_year\\":1990,\\"birth_date\\":\\"1990-03-24\\",\\"previous_id\\":\\"PsAwlRv%\\"}]"
        }
      ) {
        userErrors {
          message
        }
      }
      upsertProfileMapping(
        input: {
          advertiserId: PsAwlRv%,
          mappingSchemaV2: [{incomingKey:"user_id",destinationKey:"external_id",label:"User ID",type:STRING,isPii:false},{incomingKey:"email",destinationKey:"email",label:"Email",type:STRING,isPii:true},{incomingKey:"first_name",destinationKey:"first_name",label:"First Name",type:STRING,isPii:true},{incomingKey:"last_name",destinationKey:"last_name",label:"Last Name",type:STRING,isPii:true},{incomingKey:"phone",destinationKey:"phone",label:"Phone",type:STRING,isPii:true},{incomingKey:"address",destinationKey:"address",label:"Address",type:STRING,isPii:true},{incomingKey:"city",destinationKey:"city",label:"City",type:STRING,isPii:false},{incomingKey:"state",destinationKey:"state",label:"State",type:STRING,isPii:false},{incomingKey:"country",destinationKey:"country",label:"Country",type:STRING,isPii:false},{incomingKey:"postal_code",destinationKey:"postal_code",label:"Postal Code",type:STRING,isPii:false},{incomingKey:"timezone",destinationKey:"timezone",label:"Timezone",type:STRING,isPii:false},{incomingKey:"birth_day",destinationKey:"birth_day",label:"Birth Day",type:NUMBER,isPii:false},{incomingKey:"birth_month",destinationKey:"birth_month",label:"Birth Month",type:NUMBER,isPii:false},{incomingKey:"birth_year",destinationKey:"birth_year",label:"Birth Year",type:NUMBER,isPii:true},{incomingKey:"birth_date",destinationKey:"birth_date",label:"Birth Date",type:STRING,isPii:true}],
          isOptedIn: true,
          mappableType: "segment_io"
        }
      ) {
        userErrors {
          message
        }
      }
      
  }",
}
`;
